### PR TITLE
[IOTDB-408] Improve batch insert in partitioned storage group

### PR DESCRIPTION
**Pull request by SilverNarcissus**
Improve batch insert in partitioned storage group.
This pr is build from [IOTDB-333] and is for testing performance. 
DO NOT merge without contacting with me. Thanks~

### CODE REVIEW

File: session/src/main/java/org/apache/iotdb/session/Session.java [hunk 1]
```diff
@@ -201,6 +206,98 @@ public TSExecuteBatchStatementResp insertBatch(RowBatch rowBatch)
...
+
+    return insertSortedBatch(rowBatch);
+  }
+
+  private void sortRowBatch(RowBatch rowBatch){
+    /*
+       * following part of code sort the batch data by time,
+       * so we can insert continuous data in value list to get a better performance
+       */
+    // sort to get index, and use index to sort value list
```
Comment (jt2594838) [inline session/src/main/java/org/apache/iotdb/session/Session.java]: I would recommend that you check if the batch is already sorted before doing the actual sort.

Comment (SilverNarcissus) [reply session/src/main/java/org/apache/iotdb/session/Session.java]: fixed~

File: server/src/main/java/org/apache/iotdb/db/engine/version/SimpleFileVersionController.java [hunk 2]
```diff
@@ -31,27 +31,49 @@
...
+  private static final String FILE_PREFIX = "Version-";
   /**
    * Every time currVersion - prevVersion >= saveInterval, currVersion is persisted and prevVersion
    * is set to currVersion. When recovering from file, the version number is automatically increased
    * by saveInterval to avoid conflicts.
    */
   private static long saveInterval = 100;
-  private static final String FILE_PREFIX = "Version-";
+  /**
+   * time partition id to dividing time series into different storage group
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/engine/version/SimpleFileVersionController.java]: fixed~

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java [hunk 3]
```diff
@@ -889,27 +1126,42 @@ public void delete(String deviceId, String measurementId, long timestamp) throws
...
+      long timePartitionId = fromTimeToTimePartition(timestamp);
       // write log
       if (IoTDBDescriptor.getInstance().getConfig().isEnableWal()) {
-        if (workSequenceTsFileProcessor != null) {
-          workSequenceTsFileProcessor.getLogNode()
-              .write(new DeletePlan(timestamp, new Path(deviceId, measurementId)));
+        for (Map.Entry<Long, TsFileProcessor> entry : workSequenceTsFileProcessors.entrySet()) {
+          if (entry.getKey() <= timePartitionId) {
+            entry.getValue().getLogNode()
+                .write(new DeletePlan(timestamp, new Path(deviceId, measurementId)));
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java]: fixed~

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java [hunk 4]
```diff
@@ -534,7 +678,58 @@ private TsFileProcessor getOrCreateTsFileProcessor(boolean sequence) {
     return tsFileProcessor;
   }
 
-  private TsFileProcessor createTsFileProcessor(boolean sequence)
+  /**
+   * get processor from hashmap, flush oldest processor is necessary
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java]: fixed~

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java [hunk 5]
```diff
@@ -131,38 +134,59 @@
...
+   */
+  private long partitionIntervalForStorageGroup;
   /**
    * the schema of time series that belong this storage group
    */
   private Schema schema;
   // includes sealed and unsealed sequence TsFiles
-  private List<TsFileResource> sequenceFileList = new ArrayList<>();
-  private TsFileProcessor workSequenceTsFileProcessor = null;
+  private TreeSet<TsFileResource> sequenceFileList = new TreeSet<>(
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java]: fixed~

File: server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java [hunk 6]
```diff
@@ -439,14 +451,29 @@ public void loadNewTsFile(TsFileResource newTsFileResource)
...
   public boolean moveTsfile(File tsfileToBeMoved, File targetDir)
       throws StorageEngineException, IOException {
-    return getProcessor(tsfileToBeMoved.getParentFile().getName())
-        .moveTsfile(tsfileToBeMoved, targetDir);
+    return getProcessor(getSgByEngineFile(tsfileToBeMoved)).moveTsfile(tsfileToBeMoved, targetDir);
+  }
+
+  /**
+   * The engine file means that the file is in the engine, which is different from those external
+   * files which are not loaded.
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java]: fixed~

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBDescriptor.java [hunk 7]
```diff
@@ -299,6 +299,13 @@ private void loadProps() {
 
       conf.setDefaultTTL(Long.parseLong(properties.getProperty("default_ttl",
           String.valueOf(conf.getDefaultTTL()))));
+      // Time range for divide storage group
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/conf/IoTDBDescriptor.java]: fixed~

File: server/src/assembly/resources/conf/iotdb-engine.properties [hunk 8]
```diff
@@ -415,4 +415,17 @@ value_encoder=PLAIN
...
+compressor=SNAPPY
+
+# Time range for divide storage group
+# Time series data will divide into groups by this time range
+# Unit is second
+# default value is one day, which is 86400 seconds
+partition_interval = 86400
+
+# the num of memtables in each storage group
+# this config decide how many time series in different time partition can be insert concurrently
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply server/src/assembly/resources/conf/iotdb-engine.properties]: fixed~

File: docs/Documentation/UserGuide/3-Server/4-Config Manual.md [hunk 9]
```diff
@@ -236,6 +236,25 @@ The detail of each variables are as follows:
...
+|Type|Int64|
+|Default| 86400 |
+|Effective|Should not be changed after first start up|
+
+
+* memtable\_num\_in\_each\_storage\_group
+
+|Name| memtable\_num\_in\_each\_storage\_group |
+|:---:|:---|
+|Description| This config decide how many time series in different time partition can be insert concurrently.</br> For example, your partitionInterval is 86400 and you want to insert data in 3 different days, you should set this param >= 6 (for sequence and unsequence) |
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply docs/Documentation/UserGuide/3-Server/4-Config Manual.md]: fixed~

**Comment by SilverNarcissus** [comment]
This pr has been merged by https://github.com/apache/incubator-iotdb/pull/664. So this pr will be closed now~  If you want to review the code, please check PR664 as above

**Pull request closed without merging**